qiime2/core/type/poset.py

```diff
@@ -0,0 +1,215 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2019, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+
+def maximum_antichain(*types):
+    return POSet(*types).maximum_antichain
+
+
+def minimum_antichain(*types):
+    return POSet(*types).minimum_antichain
+
+
+class POSet:
+    class Node:
+        __le_cache = {}
+
+        def __init__(self, item):
+            self.item = item
+            self._children = {}
+            self._parents = {}
+
+        @property
+        def children(self):
+            return list(self._children)
+
+        @property
+        def parents(self):
+            return list(self._parents)
+
+        def __le__(self, other):
+            a, b = comp = self.item, other.item
+            if comp in self.__le_cache:
+                return self.__le_cache[comp]
+
+            r = a <= b
+            self.__le_cache[comp] = r
+            return r
+
+        def __lt__(self, other):
+            return self <= other and not other <= self
+
+        def __repr__(self):
+            return "<%r>" % self.item
+
+        def replace_parent(self, old, new):
+            new._children[self] = None
+            if old is not None:
+                del self._parents[old]
+            self._parents[new] = None
+
+        def replace_child(self, old, new):
+            new._parents[self] = None
+            if old is not None:
+                del self._children[old]
+            self._children[new] = None
+
+        def iter_decendents(self):
```

Comment:
```suggestion
        def iter_descendants(self):
```
There are ~20 of these tweaks in this PR. I am tagging one of each variant for easy searching in case you choose to correct spelling.